### PR TITLE
on creating a new message update LastViewedAt to now

### PR DESCRIPTION
**Pull request by Baggerone**

### CODE REVIEW

File: application/models/message.go [hunk 1]
```diff
@@ -75,6 +75,30 @@ func (m *Message) ValidateUpdate(tx *pop.Connection) (*validate.Errors, error) {
...
 
+// AfterCreate updates the LastViewedAt value on the associated ThreadParticipant to right now
+func (m *Message) AfterCreate(tx *pop.Connection) error {
+	threadP := ThreadParticipant{}
+	where := "thread_id = ? AND user_id = ?"
+	if err := DB.Where(where, m.ThreadID, m.SentByID).First(&threadP); err != nil {
+		domain.ErrLogger.Printf("error finding thread participant for Thread ID %v and User ID %v",
+			m.ThreadID, m.SentByID)
+		return nil
+	}
```
Comment (Schparky) [inline application/models/message.go]: Can use `FindByThreadIDAndUserID` here.

File: application/models/message_fixtures_test.go [hunk 2]
```diff
@@ -180,3 +184,204 @@ func Fixtures_Message_FindByUUID(ms *ModelSuite) MessageFixtures {
...
+
+	// Because of Message.AfterCreate() we need to reset the LastViewedAt attributes
+	//                     Eager User0, Lazy User1, Lazy User1, Eager User0
+	for i, tt := range []time.Time{tNow, oldTime, oldTime, tNow} {
+		if err := threadParticipants[i].UpdateLastViewedAt(tt); err != nil {
+			t.Errorf("could not update LastViewedAt ... %v", err)
+			t.FailNow()
+		}
+	}
+	println()
```
Comment (Schparky) [inline application/models/message_fixtures_test.go]:
```suggestion

```

File: application/models/message_fixtures_test.go [hunk 3]
```diff
@@ -180,3 +184,204 @@ func Fixtures_Message_FindByUUID(ms *ModelSuite) MessageFixtures {
...
+	for _, m := range messages {
+		if err := ms.DB.RawQuery(`INSERT INTO messages (content, created_at, sent_by_id, thread_id, updated_at, uuid)`+
+			`VALUES (?, ?, ?, ?, ?, ?)`,
+			m.Content, m.CreatedAt, m.SentByID, m.ThreadID, m.CreatedAt, m.Uuid).Exec(); err != nil {
+			t.Errorf("error loading messages ... %v", err)
+			t.FailNow()
+		}
+	}
+
+	// Because of Message.AfterCreate() we need to reset the LastViewedAt attributes
```
Comment (Schparky) [inline application/models/message_fixtures_test.go]: Really? I'm surprised that `RawQuery` triggers `AfterCreate`.

File: application/models/thread.go [hunk 4]
```diff
@@ -202,7 +202,9 @@ func (t *Thread) Load(fields ...string) error {
 	return nil
 }
 
-func (t *Thread) UnreadMessageCount(lastViewedAt time.Time) (int, error) {
+// UnreadMessageCount returns the number of messages on this thread that the current
+//  user has not created and for which the CreatedAt value is after the lastViewedAt value
+func (t *Thread) UnreadMessageCount(user *User, lastViewedAt time.Time) (int, error) {
```
Comment (Schparky) [inline application/models/thread.go]: I don't like that `user` is a pointer. It indicates that the function could modify it. If it's for performance reasons, I suspect that optimizations eliminate any differences.

File: application/models/message.go (hunk 1 above)
Comment (Baggerone) [reply application/models/message.go]: Good catch.

File: application/models/message_fixtures_test.go (hunk 2 above)
Comment (Baggerone) [reply application/models/message_fixtures_test.go]: Done

File: application/models/thread.go (hunk 4 above)
Comment (Baggerone) [reply application/models/thread.go]: Sounds good.

File: application/models/message_fixtures_test.go (hunk 3 above)
Comment (Baggerone) [reply application/models/message_fixtures_test.go]: Oops, I should have taken out that code after I switched from using pop to RawQuery.

**Pull request merged**